pages/recording/[id]/sourcemaps/[sourceId].tsx

```diff
@@ -0,0 +1,216 @@
+import { createHash } from "crypto";
+import { Source } from "@replayio/protocol";
+import { useRouter } from "next/router";
+import React, { useEffect, useMemo, useRef, useState } from "react";
+
+// eslint-disable-next-line no-restricted-imports
+import { client, initSocket } from "protocol/socket";
+import { assert } from "protocol/utils";
+import { setAppMode } from "ui/actions/app";
+import { ExpectedErrorScreen } from "ui/components/shared/Error";
+import LoadingScreen from "ui/components/shared/LoadingScreen";
+import { useGetRecording, useGetRecordingId } from "ui/hooks/recordings";
+import { useAppDispatch, useAppStore } from "ui/setup/hooks";
+import { ExpectedError } from "ui/state/app";
+import tokenManager from "ui/utils/tokenManager";
+
+const sha256 = createHash("sha256");
+
+type SourcemapsResult =
+  | { sourceMaps: any[]; contents: string; generatedSource: Source }
+  | {
+      error: string;
+    };
+
+async function fetchSourceMaps(filename: string, workspaceId: string) {
+  const sourceMapsResponse = await fetch(`/api/sourceMaps`, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({
+      filename,
+      workspace_id: workspaceId,
+    }),
+  });
+
+  if (sourceMapsResponse.status === 500) {
+    return { error: "Failed to fetch sourcemaps" };
+  }
+
+  const { sourceMaps } = (await sourceMapsResponse.json()) as any;
+  return sourceMaps;
+}
+
+async function loadSourceMaps(
+  recordingId: string,
+  sourceId: string,
+  workspaceId: string
+): Promise<SourcemapsResult> {
+  try {
+    console.log(`loading source maps`, { recordingId, sourceId, workspaceId });
+    const dispatchUrl =
+      new URL(location.href).searchParams.get("dispatch") || process.env.NEXT_PUBLIC_DISPATCH_URL!;
+
+    const socket = initSocket(dispatchUrl);
+    if (typeof window !== "undefined") {
+      if (window.app != null) {
+        // @ts-ignore
+        window.app.socket = socket;
+      }
+    }
+
+    const token = await tokenManager.getToken();
+    if (token.token) {
+      await client.Authentication.setAccessToken({ accessToken: token.token });
+    }
+    const { sessionId } = await client.Recording.createSession({ recordingId });
+
+    const sourcesMap: Map<string, Source> = new Map();
+    client.Debugger.addNewSourceListener(source => {
+      sourcesMap.set(source.sourceId, source);
+    });
+    client.Debugger.addNewSourcesListener(({ sources }) => {
+      for (const source of sources) {
+        sourcesMap.set(source.sourceId, source);
+      }
+    });
+    await client.Debugger.findSources({}, sessionId);
```

Comment:
this code is very similar to the source map visualizer, but i felt it was premature to extract.